docs/concepts/services/fetch.md

```diff
@@ -0,0 +1,234 @@
+    Noop,
+    ReceiveLocation(ISS),
+    FetchError(String)
+}
+
+impl Component for FetchServiceExample {
+    type Message = Msg;
+    type Properties = ();
+    fn create(_: Self::Properties, link: ComponentLink<Self>) -> Self {
+        Self {
+            ft: None,
+            iss: None,
+            link,
+            error: None,
+            fetching: false
+        }
+    }
+    fn change(&mut self, _: Self::Properties) -> bool {
+        false
+    }
+    fn update(&mut self, msg: Self::Message) -> bool {
+        match msg {
+            GetLocation => {
+                // 1. build the request
+                let request = Request::get("http://api.open-notify.org/iss-now.json")
+                    .body(Nothing)
+                    .expect("Could not build request.");
+                // 2. construct a callback
+                let callback = self.link.callback(|response: Response<Json<Result<ISS, anyhow::Error>>>| {
+                    // split up the response into data about the request's status and the returned
+                    // data from the request
+                    let (meta, Json(Ok(data))) = response.into_parts();
+                    // the condition `data.message == "success" is specific to this API which
+                    // returns a JSON object with a "message" key – not all APIs will do this! 
+                    if meta.status.is_success() && data.message == "success" {
+                        Self::Message::ReceiveLocation(match data {
+                            Ok(d) => d,
+                            Err(e) => Self::Message::FetchError(e.to_string())
+                        })
+                    } else {
+                        Self::Message::FetchError(data.message)
+                    }
+                });
+                // 3. pass the request and callback to the fetch service 
+                FetchService::fetch(request, callback);
+                self.fetching = true;
+                // we want to redraw so that the page displays a 'fetching...' message to the user 
+                // so return 'true'
+                true
+            }
+            ReceiveLocation(location) => {
+                self.iss = Some(location);
+                self.fetching = false;
+                // we want to redraw so that the page displays the location of the ISS instead of 
+                // 'fetching...'
+                true
+            }
+            FetchError(error) => {
+                self.error = error;
+                // redraw to show the error
+                true
+            }
+            _ => false
+        }
+    }
+    fn view(&self) -> Html {
+        html! {
+            <>
+                {self.is_fetching()}
+                {self.view_iss_location()}
+                {
+                    if let Some(error) = self.error {
+                        <p>{error.clone()}</p>
+                    }
+                }
+            </>
+        }
+    }
+}
```

Comment:
Turns out this example is completely broken...
```suggestion
// requires the serde and anyhow crates

use serde::Deserialize;
use yew::{
    format::{Json, Nothing},
    prelude::*,
    services::fetch::{FetchService, FetchTask, Request, Response},
};

#[derive(Deserialize, Debug, Clone)]
pub struct ISSPosition {
    latitude: String,
    longitude: String,
}

#[derive(Deserialize, Debug, Clone)]
pub struct ISS {
    message: String,
    timestamp: i32,
    iss_position: ISSPosition,
}

#[derive(Debug, Clone)]
pub enum Msg {
    GetLocation,
    ReceiveLocation(ISS),
    FetchError(String),
}

#[derive(Debug)]
pub struct FetchServiceExample {
    ft: Option<FetchTask>,
    iss: Option<ISS>,
    link: ComponentLink<Self>,
    error: Option<String>,
}
/// Some of the code to render the UI is split out into smaller functions here to make the code
/// cleaner and show some useful design patterns.
impl FetchServiceExample {
    fn view_iss_location(&self) -> Html {
        match self.iss {
            Some(ref space_station) => {
                html! {
                    <>
                        <p>{ "The ISS is at:" }</p>
                        <p>{ format!("Latitude: {}", space_station.iss_position.latitude) }</p>
                        <p>{ format!("Longitude: {}", space_station.iss_position.longitude) }</p>
                    </>
                }
            }
            None => {
                html! {
                     <button onclick=self.link.callback(|_| Msg::GetLocation)>
                         { "Where is the ISS?" }
                     </button>
                }
            }
        }
    }
    fn view_fetching(&self) -> Html {
        if self.ft.is_some() {
            html! { <p>{ "Fetching data..." }</p> }
        } else {
            html! { <p></p> }
        }
    }
    fn view_error(&self) -> Html {
        if let Some(ref error) = self.error {
            html! { <p>{ error.clone() }</p> }
        } else {
            html! {}
        }
    }
}
impl Component for FetchServiceExample {
    type Message = Msg;
    type Properties = ();

    fn create(_props: Self::Properties, link: ComponentLink<Self>) -> Self {
        Self {
            ft: None,
            iss: None,
            link,
            error: None,
        }
    }
    fn change(&mut self, _props: Self::Properties) -> bool {
        false
    }
    fn update(&mut self, msg: Self::Message) -> bool {
        use Msg::*;

        match msg {
            GetLocation => {
                // 1. build the request
                let request = Request::get("http://api.open-notify.org/iss-now.json")
                    .body(Nothing)
                    .expect("Could not build request.");
                // 2. construct a callback
                let callback =
                    self.link
                        .callback(|response: Response<Json<Result<ISS, anyhow::Error>>>| {
                            // split up the response into data about the request's status and the returned
                            // data from the request
                            let (_, Json(data)) = response.into_parts();
                            // the condition `data.message == "success" is specific to this API which
                            // returns a JSON object with a "message" key – not all APIs will do this!
                            match data {
                                Ok(d) => Msg::ReceiveLocation(d),
                                Err(e) => Msg::FetchError(e.to_string()),
                            }
                        });
                // 3. pass the request and callback to the fetch service
                let task = FetchService::fetch(request, callback).expect("failed to start request");
                // 4. store the task so it isn't cancelled immediately
                self.ft = Some(task);
                // we want to redraw so that the page displays a 'fetching...' message to the user
                // so return 'true'
                true
            }
            ReceiveLocation(location) => {
                self.iss = Some(location);
                self.ft = None;
                // we want to redraw so that the page displays the location of the ISS instead of
                // 'fetching...'
                true
            }
            FetchError(error) => {
                self.error = Some(error);
                self.ft = None;
                // redraw to show the error
                true
            }
        }
    }
    fn view(&self) -> Html {
        html! {
            <>
                { self.view_fetching() }
                { self.view_iss_location() }
                { self.view_error() }
            </>
        }
    }
}
```

Reply:
There are still a few other issues with this code but at least it compiles and runs now.

Reply:
Thanks! I really need to run it. I think it's become a bit disfigured as I've moved it around (also got a bit confused when I updated all the `FetchService` methods to be static.